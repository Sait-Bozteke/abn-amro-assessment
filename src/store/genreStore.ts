import { defineStore } from "pinia";
import { Show } from "../types/Show";

export const useGenreStore = defineStore('genreStore', {
    state: ()=>({
        genres : ['Drama', 'Comedy', 'Action']
    })
})