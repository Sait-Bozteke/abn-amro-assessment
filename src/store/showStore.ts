import { defineStore } from 'pinia';
import { ref } from 'vue';
import {
  getAllShows,
  getShowById,
  searchShowByName,
} from '../services/showService';
import { Show } from '../types/Show';

export const useShowStore = defineStore('showStore', () => {
  const shows = ref<Show[]>([]);
  const selectedShow = ref<Show | null>(null);
  const loading = ref(false);
  const error = ref<string | null>(null);

  const fetchAllShows = async () => {
    loading.value = true;
    error.value = null;
    try {
      shows.value = await getAllShows();
    } catch {
      error.value = 'Failed to fetch shows';
    } finally {
      loading.value = false;
    }
  };

  const fetchShowById = async (id: number) => {
    loading.value = true;
    error.value = null;
    try {
      selectedShow.value = await getShowById(id);
    } catch (err) {
      console.error(err);
      error.value = 'Failed to fetch show details';
    } finally {
      loading.value = false;
    }
  };

  const searchShows = async (query: string) => {
    loading.value = true;
    error.value = null;
    try {
      shows.value = await searchShowByName(query);
    } catch {
      error.value = 'Failed to search shows';
    } finally {
      loading.value = false;
    }
  };

  return {
    shows,
    selectedShow,
    loading,
    error,
    fetchAllShows,
    fetchShowById,
    searchShows,
  };
});
