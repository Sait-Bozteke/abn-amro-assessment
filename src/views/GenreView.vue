<template>
  <div class="container my-4">
    <h2 class="mb-4">{{ genre }} Shows</h2>

    <div class="input-group mb-4">
      <input
        v-model="searchQuery"
        type="text"
        class="form-control"
        placeholder="Search for shows"
      />
      <button class="btn btn-primary">Search</button>
    </div>

    <div v-if="loading" class="d-flex justify-content-center mb-4">
      <div class="spinner-border text-primary" role="status">
        <span class="visually-hidden">Loading...</span>
      </div>
    </div>

    <div v-if="error" class="alert alert-danger">
      {{ error }}
    </div>

    <div class="row">
      <ShowCard
        v-for="show in filteredShows"
        :key="show.id"
        :show="show"
        class="col-md-4 mb-4"
      />
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, onMounted, ref, computed } from 'vue';
import { useRoute } from 'vue-router';
import ShowCard from '../components/ShowCard.vue';
import { useShowStore } from '../store/showStore';

export default defineComponent({
  components: {
    ShowCard,
  },
  setup() {
    const showStore = useShowStore();
    const route = useRoute();
    const genre = route.params.genre as string;

    const searchQuery = ref('');

    const filteredShows = computed(() => {
      const query = searchQuery.value.toLowerCase();
      return showStore.shows.filter(
        (show) =>
          show.genres.includes(genre) &&
          show.name.toLowerCase().includes(query),
      );
    });

    onMounted(() => {
      showStore.fetchAllShows();
    });

    return {
      genre,
      searchQuery,
      filteredShows,
      loading: showStore.loading,
      error: showStore.error,
    };
  },
});
</script>

<style scoped>
.container {
  margin-top: 20px;
}

.input-group {
  max-width: 500px;
  margin: 0 auto 20px auto;
}

.input-group input {
  border-radius: 0.25rem 0 0 0.25rem;
}

.input-group button {
  border-radius: 0 0.25rem 0.25rem 0;
}
</style>
