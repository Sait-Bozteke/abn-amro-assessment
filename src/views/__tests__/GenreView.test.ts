import { mount } from '@vue/test-utils';
import GenreView from '../GenreView.vue';
import { useShowStore } from '../../store/showStore';
import { createRouter, createMemoryHistory, RouteRecordRaw } from 'vue-router';

jest.mock('../../store/showStore', () => ({
  useShowStore: jest.fn().mockReturnValue({
    shows: [
      { id: 1, name: 'Breaking Bad', genres: ['Drama', 'Crime'] },
      { id: 2, name: 'Friends', genres: ['Comedy'] },
    ],
    fetchAllShows: jest.fn(),
    loading: false,
    error: null,
  }),
}));

const routes: RouteRecordRaw[] = [
  { path: '/genre/:genre', name: 'GenreView', component: GenreView },
];

const router = createRouter({
  history: createMemoryHistory(),
  routes,
});

describe('GenreView.vue', () => {
  it('renders correctly and matches snapshot', async () => {
    const wrapper = mount(GenreView, {
      global: {
        plugins: [router],
      },
      props: {
        genre: 'Drama',
      },
    });

    await router.isReady();
    await wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('displays loading spinner while fetching shows', async () => {
    (useShowStore as unknown as jest.Mock).mockReturnValue({
      shows: [],
      fetchAllShows: jest.fn(),
      loading: true,
      error: null,
    });

    const wrapper = mount(GenreView, {
      global: {
        plugins: [router],
      },
      props: {
        genre: 'Drama',
      },
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.find('.spinner-border').exists()).toBe(true);
  });

  it('displays error message if there is an error', async () => {
    (useShowStore as unknown as jest.Mock).mockReturnValue({
      shows: [],
      fetchAllShows: jest.fn(),
      loading: false,
      error: 'Something went wrong',
    });

    const wrapper = mount(GenreView, {
      global: {
        plugins: [router],
      },
      props: {
        genre: 'Drama',
      },
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.find('.alert-danger').text()).toBe('Something went wrong');
  });
});
