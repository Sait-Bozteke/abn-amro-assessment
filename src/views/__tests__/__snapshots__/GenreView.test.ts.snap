// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenreView.vue renders correctly and matches snapshot 1`] = `
"<div class="container my-4" genre="Drama">
  <h2 class="mb-4"> Shows</h2>
  <div class="input-group mb-4"><input type="text" class="form-control" placeholder="Search for shows"><button class="btn btn-primary">Search</button></div>
  <!--v-if-->
  <!--v-if-->
  <div class="row"></div>
</div>"
`;
